README.md
# ftrack-sftp-accessor

An ftrack sftp [accessor](https://ftrack-python-api.readthedocs.io/en/backlog-threading-story/locations/overview.html?highlight=accessor#accessors). 


**Note**
> To see paths resolved in Studio, the [resolver listener](https://github.com/ftrackhq/ftrack-connect/blob/main/resource/hook/resolver.py) need to be running, either through [ftrack-connect](https://github.com/ftrackhq/ftrack-connect) or manually.

# Installation

Install using pip:

    $ pip install ftrack-sftp-accessor

Install from sources:

    $ git clone https://github.com/ftrackhq/ftrack-sftp-accessor.git
    $ pip install poetry
    $ poetry build
    $ pip install dist/ftrack_sftp_accessor-<VERSION>-py3-none-any.whl


## Documentation

Full documentation, including installation and setup guides can be found at https://ftrack-sftp-accessor.readthedocs.io/en/latest/

## Usage

The main plugin can be found in the plugins folder. This folder may be registered using the FTRACK_EVENT_PLUGIN_PATH ftrack environment variable so that it is picked up when ftrack is started.

Examples of how to use the plugin can be found in the scripts folder. The simplest way to launch ftrack with the accessor is scripts/start_ftrack_with_sftp.py. 

# Contributing

Please refer to the [CONTRIBUTING.md](./CONTRIBUTING.md) 

docs/release_notes.rst
..
    :copyright: Copyright (c) 2014 ftrack

.. _release/release_notes:

*************
Release Notes
*************

.. release:: 0.5.1
    :date: 2023-02-28

    .. change:: changed

        Improve toml file with missing details for pypi release.

    .. change:: changed

        ftrack-action-handler is now optional.

.. release:: 0.5.0
    :date: 2023-02-16

    .. change:: changed

        Improve toml file with missing information.

    .. change:: added

        Initial code implementation.



tests/test_ftrack_sftp_accessor.py
from ftrack_sftp_accessor import __version__


def test_version():
    assert __version__ == "0.1.0"


scripts/sftp_upload.py
# :coding: utf-8
# :copyright: Copyright (c) 2014-2022 ftrack

import ftrack_api.structure.id
from ftrack_sftp_accessor.sftp import SFTPAccessor
import os


session = ftrack_api.Session()
location = session.query('Location where name is "studio.sftp"').one()

hostname = os.getenv("FTRACK_SFTP_ACCESSOR_HOSTNAME", None)
username = os.getenv("FTRACK_SFTP_ACCESSOR_USERNAME", None)
port = os.getenv("FTRACK_SFTP_ACCESSOR_PORT", 22)
folder = os.getenv("FTRACK_SFTP_ACCESSOR_FOLDER", "ftrack")

# Setup accessor
location.accessor = SFTPAccessor(hostname, username, port=port, folder=folder)
location.structure = ftrack_api.structure.standard.StandardStructure()
location.priority = 30

version = session.get("AssetVersion", "83b88aa2-b739-4ed1-b840-2ccdac7da271")
version.create_component(
    path="/Users/ian/Downloads/film0001_512kb.mp4", location=location
)


scripts/sftp_download.py
# :coding: utf-8
# :copyright: Copyright (c) 2014-2022 ftrack

import ftrack_api.structure.id
import ftrack_api.accessor.disk
import os

from ftrack_sftp_accessor.sftp import SFTPAccessor

"""
Copies a component from a sftp location to a local location
"""
session = ftrack_api.Session()
sftp_location = session.ensure("Location", {"name": "studio.sftp"})

hostname = os.getenv("FTRACK_SFTP_ACCESSOR_HOSTNAME", None)
username = os.getenv("FTRACK_SFTP_ACCESSOR_USERNAME", None)
port = os.getenv("FTRACK_SFTP_ACCESSOR_PORT", 22)
folder = os.getenv("FTRACK_SFTP_ACCESSOR_FOLDER", "ftrack")

sftp_location.accessor = SFTPAccessor(hostname, username, port=port, folder=folder)
sftp_location.structure = ftrack_api.structure.standard.StandardStructure()
sftp_location.priority = 30

ftrack_location = session.ensure("Location", {"name": "ftrack.connect"})
ftrack_location.structure = ftrack_api.structure.standard.StandardStructure()
ftrack_location.accessor = ftrack_api.accessor.disk.DiskAccessor(
    prefix="/Users/ian/ftrack_storage"
)

version = session.get("AssetVersion", "9d5d86b7-46b5-49fc-993c-03ce885a6950")

for component in version["components"]:
    if component["name"] == "main":
        ftrack_location.add_component(component, sftp_location)


ftrack_sftp_accessor/sftp.py
# :coding: utf-8
# :copyright: Copyright (c) 2014-2022 ftrack

import os
import logging
import paramiko

from ftrack_api.accessor.base import Accessor
from ftrack_api.data import FileWrapper
from ftrack_api.exception import (
    AccessorOperationFailedError,
    AccessorUnsupportedOperationError,
    AccessorResourceInvalidError,
    AccessorResourceNotFoundError,
    AccessorContainerNotEmptyError,
    AccessorParentResourceNotFoundError,
)


class SFTPAccessor(Accessor):
    """Provide SFTP location access."""

    def __init__(self, host, username, port=22, password=None, folder=None):
        """Initialise location accessor.

        Uses the server credentials specified by *host*, *password*, *port* and *password*
        to create a sftp connection.

        If specified, *folder* indicates the subfolder where assets are stored
        """
        self._host = host
        self._username = username
        self._password = password
        self._port = port
        self._folder = folder
        self._sftp = None
        self._ssh = None
        self._logger = logging.getLogger(__name__ + "." + self.__class__.__name__)
        super(SFTPAccessor, self).__init__()

    def __deepcopy__(self, memo):
        """Return a new SFTPAccessor instance"""
        return SFTPAccessor(
            self._host, self._username, self._port, self._password, self._folder
        )

    @property
    def ssh(self):
        """Return SSH resource."""
        if self._ssh is None:
            self._logger.debug("Initialising SSH Session")
            self._ssh = paramiko.SSHClient()
            self._ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
            if not self._password:
                # attempt to connect using keyfile if no password specified
                self._ssh.connect(
                    self._host,
                    port=self._port,
                )
            else:
                self._ssh.connect(
                    hostname=self._host,
                    port=self._port,
                    username=self._username,
                    password=self._password,
                )
        return self._ssh

    @property
    def sftp(self):
        """Return SFTP resource."""
        self._logger.debug("Initialising SFTP Session")
        if self._sftp is None:
            self._sftp = self.ssh.open_sftp()

            if self._folder is not None:
                self._sftp.chdir(self._folder)

        return self._sftp

    def list(self, resource_identifier):
        """Return list of entries in *resource_identifier* container.

        Each entry in the returned list should be a valid resource identifier.

        Raise :py:class:`~ftrack.ftrackerror.AccessorResourceNotFoundError` if
        *resource_identifier* does not exist or
        :py:class:`~ftrack.ftrackerror.AccessorResourceInvalidError` if
        *resource_identifier* is not a container.

        """
        if not resource_identifier.endswith("/"):
            resource_identifier += "/"

        if resource_identifier == "/":
            resource_identifier = ""

        try:
            self.sftp.stat(resource_identifier)
            return self.sftp.listdir(resource_identifier)
        except IOError:
            self._logger.debug(
                f"Returning an empty list as resource identifier {resource_identifier} doesn't exist"
            )
            return []

    def exists(self, resource_identifier):
        """Return if *resource_identifier* is valid and exists in location."""
        # Root directory always exists
        if not resource_identifier:
            return True

        return self.is_container(resource_identifier) or self.is_file(
            resource_identifier
        )

    def is_file(self, resource_identifier):
        """Return whether *resource_identifier* refers to a file."""
        # Root is a directory
        if not resource_identifier:
            return False

        resource_identifier = resource_identifier.rstrip("/")

        try:
            file_object = self.sftp.stat(resource_identifier)
        except IOError:
            self._logger.debug(
                f"Returning is not file as resource identifier {resource_identifier} doesn't exist"
            )
            file_object = None
        except Exception as error:
            raise AccessorOperationFailedError(
                operation="is_file",
                resource_identifier=resource_identifier,
                error=error,
            )

        return file_object is not None

    def is_container(self, resource_identifier):
        """Return whether *resource_identifier* refers to a container."""
        # Root is a directory
        if not resource_identifier:
            return True

        file_objects = self.list(resource_identifier)

        try:
            next(iter(file_objects))
        except StopIteration:
            return False
        else:
            return True

    def is_sequence(self, resource_identifier):
        """Return whether *resource_identifier* refers to a file sequence."""
        raise AccessorUnsupportedOperationError("is_sequence")

    def open(self, resource_identifier, mode="rb"):
        """Return :py:class:`~ftrack.Data` for *resource_identifier*."""
        if self.is_container(resource_identifier):
            raise AccessorResourceInvalidError(
                resource_identifier=resource_identifier,
                message="Cannot open a directory: {resource_identifier}",
            )

        try:
            file_obj = self.sftp.open(resource_identifier, mode)
        except IOError:
            self._logger.debug(f"Creating SFTP Resource {resource_identifier}")
            if "w" not in mode and "a" not in mode:
                raise AccessorResourceNotFoundError(
                    resource_identifier=resource_identifier
                )

            self.ssh.exec_command(f"touch {resource_identifier}")
            file_obj = self.sftp.open(resource_identifier, mode)

        if "w" in mode:
            file_obj = self.sftp.open(resource_identifier, mode)

        return FileWrapper(file_obj)

    def remove(self, resource_identifier):
        """Remove *resource_identifier*.

        Raise :py:class:`~ftrack.ftrackerror.AccessorResourceNotFoundError` if
        *resource_identifier* does not exist.

        """
        self._logger.debug(f"Removing SFTP Resource {resource_identifier}")
        if self.is_file(resource_identifier):
            self.sftp.remove(resource_identifier)

        elif self.is_container(resource_identifier):
            contents = self.list(resource_identifier)
            if contents:
                raise AccessorContainerNotEmptyError(
                    resource_identifier=resource_identifier
                )

            self.sftp.remove(resource_identifier + "/")

        else:
            raise AccessorResourceNotFoundError(resource_identifier=resource_identifier)

    def get_container(self, resource_identifier):
        """Return resource_identifier of container for *resource_identifier*.

        Raise
        :py:class:`~ftrack.ftrackerror.AccessorParentResourceNotFoundError` if
        container of *resource_identifier* could not be determined.

        """
        if os.path.normpath(resource_identifier) in ("/", ""):
            raise AccessorParentResourceNotFoundError(
                resource_identifier=resource_identifier,
                message="Could not determine container for "
                "{resource_identifier} as it is the root.",
            )

        return os.path.dirname(resource_identifier.rstrip("/"))

    def make_container(self, resource_identifier, recursive=True):
        """Make a container at *resource_identifier*.

        If *recursive* is True, also make any intermediate containers.

        """
        if not resource_identifier:
            # Root directory
            return

        if not resource_identifier.endswith("/"):
            resource_identifier += "/"

        if self.exists(resource_identifier):
            if self.is_file(resource_identifier):
                raise AccessorResourceInvalidError(
                    resource_identifier=resource_identifier,
                    message=("Resource {resource_identifier} already exists as a file"),
                )

            else:
                return

        parent = self.get_container(resource_identifier)

        if not self.is_container(parent):
            if recursive:
                self.make_container(parent, recursive=recursive)
            else:
                raise AccessorParentResourceNotFoundError(resource_identifier=parent)

        self.sftp.mkdir(resource_identifier)

    def get_url(self, resource_identifier=None):
        """Return url for *resource_identifier*."""

        if self._folder:
            return (
                f"sftp://{self._host}:{self._port}/{self._folder}/{resource_identifier}"
            )
        return f"sftp://{self._host}:{self._port}/{resource_identifier}"


plugins/sftp_location.py
# :coding: utf-8
# :copyright: Copyright (c) 2014-2022 ftrack

import ftrack_api
import ftrack_api.entity.location
import ftrack_api.accessor.disk
import ftrack_api.structure.id
import logging
import os
import sys
import functools

NAME = "ftrack-sftp-accessor"
VERSION = "0.1.0"

logger = logging.getLogger("{}".format(NAME.replace("-", "_")))


def configure_location(event):
    from ftrack_sftp_accessor.sftp import SFTPAccessor

    """Configure locations for session."""
    session = event["data"]["session"]

    # Find location(s) and customise instances.
    location = session.query('Location where name is "studio.sftp"').one()

    hostname = os.getenv("FTRACK_SFTP_ACCESSOR_HOSTNAME", None)
    port = os.getenv("FTRACK_SFTP_ACCESSOR_PORT", 22)
    username = os.getenv("FTRACK_SFTP_ACCESSOR_USERNAME", None)
    password = os.getenv("FTRACK_SFTP_ACCESSOR_PASSWORD", None)
    folder = os.getenv("FTRACK_SFTP_ACCESSOR_FOLDER", None)

    # Setup accessor to use bucket
    location.accessor = SFTPAccessor(
        hostname, username, port=port, password=password, folder=folder
    )
    location.structure = ftrack_api.structure.standard.StandardStructure()
    location.priority = 30

    # Setup any other locations you require
    ftrack_location = session.query('Location where name is "ftrack.connect"').one()
    ftrack_location.structure = ftrack_api.structure.standard.StandardStructure()
    ftrack_location.accessor = ftrack_api.accessor.disk.DiskAccessor(
        prefix="/Users/ian/ftrack_storage"
    )


def register(session):
    if not isinstance(session, ftrack_api.session.Session):
        return

    logging.info("Registering sftp accessor")

    session.event_hub.subscribe(
        "topic=ftrack.api.session.configure-location and source.user.username={}".format(
            session.api_user
        ),
        configure_location,
    )
